p2p/transport_mconn.go

```diff
@@ -0,0 +1,646 @@
+
+	maxIncomingConnections int
+	dialTimeout            time.Duration
+	handshakeTimeout       time.Duration
+	filterTimeout          time.Duration
+
+	logger   log.Logger
+	listener net.Listener
+
+	closeOnce sync.Once
+	chAccept  chan *mConnConnection
+	chError   chan error
+	chClose   chan struct{}
+
+	// FIXME This is a vestige from the old transport, and should be managed
+	// by the router once we rewrite the P2P core.
+	conns       ConnSet
+	connFilters []ConnFilterFunc
+}
+
+// NewMConnTransport sets up a new MConn transport.
+func NewMConnTransport(
+	logger log.Logger,
+	nodeInfo NodeInfo, // FIXME should use DefaultNodeInfo, left for code compatibility
+	privKey crypto.PrivKey,
+	mConnConfig conn.MConnConfig,
+	opts ...MConnTransportOption,
+) *MConnTransport {
+	m := &MConnTransport{
+		privKey:      privKey,
+		nodeInfo:     nodeInfo.(DefaultNodeInfo),
+		mConnConfig:  mConnConfig,
+		channelDescs: []*ChannelDescriptor{},
+
+		dialTimeout:      defaultDialTimeout,
+		handshakeTimeout: defaultHandshakeTimeout,
+		filterTimeout:    defaultFilterTimeout,
+
+		logger:   logger,
+		chAccept: make(chan *mConnConnection),
+		chError:  make(chan error),
+		chClose:  make(chan struct{}),
+
+		conns:       NewConnSet(),
+		connFilters: []ConnFilterFunc{},
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+	return m
+}
+
+// SetChannelDescriptors implements Transport.
+func (m *MConnTransport) SetChannelDescriptors(chDescs []*conn.ChannelDescriptor) {
+	m.channelDescs = chDescs
+}
+
+// Listen asynchronously listens for inbound connections on the given endpoint.
+// It must be called exactly once before calling Accept(), and the caller must
+// call Close() to shut down the listener.
+func (m *MConnTransport) Listen(endpoint Endpoint) error {
+	if m.listener != nil {
+		return errors.New("MConn transport is already listening")
+	}
+	err := m.normalizeEndpoint(&endpoint)
+	if err != nil {
+		return fmt.Errorf("invalid MConn listen endpoint %q: %w", endpoint, err)
+	}
+
+	m.listener, err = net.Listen("tcp", fmt.Sprintf("%v:%v", endpoint.IP, endpoint.Port))
+	if err != nil {
+		return err
+	}
+	if m.maxIncomingConnections > 0 {
+		m.listener = netutil.LimitListener(m.listener, m.maxIncomingConnections)
+	}
+
+	// Spawn a goroutine to accept inbound connections asynchronously.
+	go m.accept()
+
+	return nil
+}
+
+// accept accepts inbound connections in a loop, and asynchronously handshakes
+// with the peer to avoid head-of-line blocking. Established connections are
+// passed to Accept() via the channel m.chAccept.
+// See: https://github.com/tendermint/tendermint/issues/204
+func (m *MConnTransport) accept() {
+	for {
+		tcpConn, err := m.listener.Accept()
+		if err != nil {
+			select {
+			case m.chError <- err:
+			case <-m.chClose:
+			}
+			return
+		}
+		go func() {
+			err := m.filterTCPConn(tcpConn)
+			if err != nil {
+				_ = tcpConn.Close()
+				select {
+				case m.chError <- err:
+				case <-m.chClose:
+				}
```

Comment:
This block is called three times. Perhaps this warrants an auxiliary function that accepts an `error` argument.

Reply:
I don't know. I think it's easier to read when all the logic is in the same place, and the different branches do different things. Having to jump around to get the full picture makes it easier to miss subtle details.

Besides, some of this stuff will be removed/simplified once we move the TCP connection filtering and such out of the transport and into the router/switch.

Reply:
> Besides, some of this stuff will be removed/simplified once we move the connection filtering out of the transport and into the router/switch.

That's fine. But in this case we're calling the exact same branch of logic three times so typically when I see that I judge that it probably warrants a method. Not a blocker though. Feel free to resolve.

Reply:
I think it's a judgement call, but in this specific case I think it's useful to optimize for understandability over deduplication.